### PR TITLE
network: retransmit stale transactions

### LINKED ISSUE #1422: Transaction retransmission mechanism
We need it for Neo 2 mostly, although Neo 3 already has support for block index data for its mempool items (this needs to be ported over to Neo 2 to then leverage it for block rebroadcasting). Still, I hope Neo 3 won't need it long-term because of future tail-cutting changes.

The idea is that in `RemoveStale` loop we should also check for tx block vs current block and if that is higher than a certain threshold (`2×ValidatorsCount`, but no less than 5) we should add it into some separate slice and after the loop send this slice to another goroutine that would then push these transactions into `network.Server` broadcasting loop.

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1422 .

### CODE REVIEW

**Review by AnnaShaleva — APPROVED** [review]

File: pkg/core/mempool/mem_pool.go [hunk 1]
```diff
@@ -329,6 +351,19 @@ func (mp *Pool) GetVerifiedTransactions() []TxWithFee {
 	return t
 }
 
+// GetStaleTransactions returns transactions which are in Pool for more
+// than StaleThreshold blocks.
+func (mp *Pool) GetStaleTransactions() []*transaction.Transaction {
+	mp.lock.Lock()
```
Comment (roman-khimov) [inline pkg/core/mempool/mem_pool.go:357]: RLock

File: pkg/network/server.go [hunk 2]
```diff
@@ -968,6 +969,38 @@ func (s *Server) broadcastTxHashes(hs []util.Uint256) {
 	})
 }
 
+// broadcastStaleTxLoop is a loop for rebroadcasting tx which are in a mempool
+// for too long.
+func (s *Server) broadcastStaleTxLoop() {
```
Comment (roman-khimov) [inline pkg/network/server.go:974]: And I don't think we need this. We only need `Server` to give `s.transactions` channel to the mempool (maybe along with `resendThreshold`).

File: pkg/core/mempool/mem_pool.go [hunk 3]
```diff
@@ -287,6 +300,7 @@ func (mp *Pool) RemoveStale(isOK func(*transaction.Transaction) bool) {
 	sort.Slice(newClaims, func(i, j int) bool {
 		return newClaims[i].Cmp(newClaims[j]) < 0
 	})
+	mp.staleTxs = staleTxs
```
Comment (roman-khimov) [inline pkg/core/mempool/mem_pool.go:303]: And here you'll just do
```
go resendTransactions(staleTxs)
```
That will send them into the goroutine provided earlier.

File: pkg/core/mempool/mem_pool.go [hunk 4]
```diff
@@ -49,6 +50,9 @@ type Pool struct {
 	inputs       []*transaction.Input
 	claims       []*transaction.Input
 
+	resendThreshold uint32
+	staleTxs        []*transaction.Transaction
```
Comment (roman-khimov) [inline pkg/core/mempool/mem_pool.go:54]: And you won't gonna need `staleTxs` here at all.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/core/mempool/mem_pool.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/mempool/mem_pool.go:357]: Fixed.

File: pkg/core/mempool/mem_pool.go [hunk 5]
```diff
@@ -159,9 +163,14 @@ func dropInputFromSortedSlice(slice *[]*transaction.Input, input *transaction.In
 }
 
 // Add tries to add given transaction to the Pool.
-func (mp *Pool) Add(t *transaction.Transaction, fee Feer) error {
+func (mp *Pool) Add(t *transaction.Transaction, fee Feer, hs ...uint32) error {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/mempool/mem_pool.go:166]: Fixed.

File: pkg/network/server.go (hunk 2 above)
Comment (fyrchik) [reply pkg/network/server.go:974]: Fixed.

File: pkg/core/mempool/mem_pool.go [hunk 6]
```diff
@@ -277,6 +287,9 @@ func (mp *Pool) RemoveStale(isOK func(*transaction.Transaction) bool) {
 					newClaims = append(newClaims, &claim.Claims[i])
 				}
 			}
+			if mp.resendThreshold != 0 && itm.pooledAt+mp.resendThreshold <= height {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/mempool/mem_pool.go:290]: Fixed.

File: pkg/core/mempool/mem_pool.go (hunk 3 above)
Comment (fyrchik) [reply pkg/core/mempool/mem_pool.go:303]: Fixed.

File: pkg/core/mempool/mem_pool.go (hunk 4 above)
Comment (fyrchik) [reply pkg/core/mempool/mem_pool.go:54]: Fixed.

File: pkg/core/mempool/mem_pool.go [hunk 7]
```diff
@@ -26,6 +27,7 @@ var (
 // item represents a transaction in the the Memory pool.
 type item struct {
 	txn        *transaction.Transaction
+	blockStamp uint32
 	timeStamp  time.Time
```
Comment (roman-khimov) [inline pkg/core/mempool/mem_pool.go:31]: Just drop it. Or maybe even just backport 5df726db68afc87bfbac11194b3f925ce4efd2fd.

File: pkg/network/server.go [hunk 8]
```diff
@@ -177,6 +177,7 @@ func (s *Server) Start(errChan chan error) {
 
 	s.tryStartConsensus()
 
+	go s.broadcastStaleTxLoop()
```
Comment (roman-khimov) [inline pkg/network/server.go:180]: Don't `go` here, it's just not needed any more.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/network/server.go (hunk 8 above)
Comment (roman-khimov) [reply pkg/network/server.go:180]: And there is no `Loop` inside anyway.

File: pkg/core/mempool/mem_pool.go (hunk 7 above)
Comment (fyrchik) [reply pkg/core/mempool/mem_pool.go:31]: Fixed.

File: pkg/network/server.go (hunk 8 above)
Comment (fyrchik) [reply pkg/network/server.go:180]: Fixed.

**Review by roman-khimov — APPROVED** [review]
And now it should be easy to forward-port it to master.

**Pull request merged**